ensure scheduler preemptor behaves in an efficient/correct path

Huang-Wei opened the pull request:
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

- don't update nominatedMap cache when Pop() an element from activeQ
- instead, delete the nominated info from cache when it's bound (scheduled)

**Which issue(s) this PR fixes**:

Fixes #70622

**Special notes for your reviewer**:

The most significant change introduced in this PR is: when popping a pod, scheduler doesn't update internal cache from nominatedMap immediately. Instead, it invalidates the cache until the pod is bound.

Why this? It's because in a very rare case: when a high priority pod comes in, and it's unschedulable (failed in scheduling) (1), it got a chance to try "preemption" (2) and preempt low priority pods (3) to make room.

During phase (1), in function Error() (1.1), it's put back into unschedulableQ where cache nominatedMap is being re-updated, the key point here is: the function is asynchronous (in a goroutine). In other words, after (3) is finished, a backfill pod for the preempted low priority pod (suppose it's managed by a deployment/replicaset) can be spawned and come into scheduling cycle, and it happens prior to (1.1). At this moment, it doesn't know a Nominated pod has been there (as cache hasn't been re-updated), then it's created and enters running state, but it will definitely be preempted again. So this case can happen again and again, although not endless, but really wastes resources to do unnecessary scheduling/preemption.

Along with this PR, I wrote an e2e test to simulate the issue.

**Does this PR introduce a user-facing change?**:

```release-note
Fix a potential bug that scheduler preempts unnecessary pods.
```

/sig scheduling

resouer commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -219,7 +219,7 @@ func (p *PriorityQueue) addNominatedPodIfNeeded(pod *v1.Pod) {
 		for _, np := range p.nominatedPods[nnn] {
 			if np.UID == pod.UID {
-				klog.Errorf("Pod %v/%v already exists in the nominated map!", pod.Namespace, pod.Name)
+				klog.Infof("Pod %v/%v already exists in the nominated map!", pod.Namespace, pod.Name)
```
Downgrade to `Infof` seems a bit aggressive, as it seems this case is unexpected.

Huang-Wei replied:
@resouer Thanks for the comment. Based on current solution, the message occurs more often and it's as design. So I'd put your comment aside for a while, until we come to the consensus on core idea of this PR.

Huang-Wei commented:
CI is green now (including the e2e test in this PR).

/cc @bsalamat @resouer @k82cn @ravisantoshgudimetla

Huang-Wei commented:
/priority important-soon

Huang-Wei commented on pkg/scheduler/core/generic_scheduler.go:
```diff
@@ -511,7 +511,7 @@ func addNominatedPods(podPriority int32, meta algorithm.PredicateMetadata,
 	nodeInfoOut := nodeInfo.Clone()
 	for _, p := range nominatedPods {
-		if util.GetPodPriority(p) >= podPriority {
+		if util.GetPodPriority(p) >= util.GetPodPriority(pod) && p.UID != pod.UID {
```
checking UID is necessary b/c right now `pod` itself exists in nominatedPods.

Huang-Wei commented on pkg/scheduler/internal/queue/scheduling_queue_test.go:
```diff
@@ -112,8 +112,8 @@ func TestPriorityQueue_Add(t *testing.T) {
 	}
-	if len(q.nominatedPods) != 0 {
-		t.Errorf("Expected nomindatePods to be empty: %v", q.nominatedPods)
+	if len(q.nominatedPods["node1"]) != 2 {
```
it's a necessary change as right now popping out a pod won't clear cache from nominatedPods cache.

Huang-Wei commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -411,13 +411,18 @@ func (p *PriorityQueue) Delete(pod *v1.Pod) error {
 // may make pending pods with matching affinity terms schedulable.
 func (p *PriorityQueue) AssignedPodUpdated(pod *v1.Pod) {
+	p.lock.Lock()
+	p.deleteNominatedPodIfExists(pod)
```
The reason why I didn't put the logic in `assume()` is current assume design has no interaction with scheduling_queue, if we have to do that, it will result in pretty a lot of interface changes.

I don't quite understand "double counting". Could you collaborate a bit?

Based on current solution, in between assume() and bound(), the nominatedMap is still kept, so it's still calculated when scheduling low priority pending pods. But the interval would be pretty short.

bsalamat replied:
"double counting" comes from the fact that an assumed pod is added to the NodeInfo in the scheduler cache and its resources are deducted from the node. It also remains in the "NominatedNode" map. So, when scheduler tries scheduling another pod, it adds the resources needed by the "nominated pods" again to the resources used on the node. So, it may determine the incoming pod unschedulable due to lack of resources, while there may be enough resources.

The easiest way to fix it is probably adding a pointer of the scheduling queue to `factory.Config`. It will then be available to the `scheduler.Scheduler`. Then you can expose and call `deleteNominatedPodIfExists` form `scheduler.go -> assum()`.

Huang-Wei replied:
Ah I see the point. Thanks for the explanation.

Will put logic into assume().

Huang-Wei commented on test/e2e/scheduling/preemption.go:
```diff
@@ -353,3 +358,218 @@ var _ = SIGDescribe("PodPriorityResolution [Serial]", func() {
+			nodeCopy.ResourceVersion = "0"
+			delete(nodeCopy.Status.Capacity, fakecpu)
+			_, err := cs.CoreV1().Nodes().UpdateStatus(nodeCopy)
+			framework.ExpectNoError(err)
```
I had the same concern. But I see some other e2e (test/e2e/scheduling/preemption.go) also leave the Priority there, without cleaning.

bsalamat replied:
We should clean up those too. Not in this PR though.

Huang-Wei replied:
Sounds good.

Huang-Wei commented on test/e2e/scheduling/preemption.go:
```diff
@@ -353,3 +358,218 @@ var _ = SIGDescribe("PodPriorityResolution [Serial]", func() {
+		framework.Logf("length of pods created so far: %v", len(podEvents))
+		// 13 = 5+4+4, which implies ReplicaSet{1,2,3} should create and only create
+		// exactly the "replicas" of pods
+		if podEventsNum := len(podEvents); podEventsNum != 13 {
```
I like this neat case, and it looks quite similar like:

https://github.com/kubernetes/kubernetes/blob/12402742041386b82c89b1c64cb855bdc1691928/test/integration/scheduler/preemption_test.go#L395

Current case does't check if eventual 5 pending pods matches the name of the 5 pending ones created in the beginning.

bsalamat replied:
You are right. That integration test actually exists. Have you tried running it with "stress" to see if it fails at all?

Huang-Wei commented:
@bsalamat I've updated the logic of deleting nominatedPod from cache to `assume()` - I didn't expose the whole scheduleQueue in factor.Config, instead expose a function. PTAL (latest commit).

Regarding the test, I'm still trying to build an **_integration test_**, but no luck to reproduce the issue so far.

bsalamat commented on pkg/scheduler/factory/factory.go:
```diff
@@ -135,6 +135,10 @@ type Config struct {
+
+	// DeleteNominatedPodIfExists is called when a pod is assumed
+	// It will delete the pod from internal cache if it's a nominated pod
+	DeleteNominatedPodIfExists func(pod *v1.Pod)
```
I would still prefer exposing the scheduling queue here. This function looks too low level to me to get exposed in the scheduler config. We have the scheduler cache, event recorder, volume binder, etc. here. The scheduling queue looks more of the same nature of the existing members of this struct.

bsalamat reviewed (commented):
@Huang-Wei if you can verify in your real clusters that the fix actually works, I am fine with having the fix in 1.13 and adding the test in the next few days.

Huang-Wei commented:
@bsalamat yes, I can reproduce both manually and using the e2e test I wrote. Could you lable this PR milestone v1.13?

> I am fine with having the fix in 1.13 and adding the test in the next few days.

This SGTM. I will update this PR to remove the e2e test and address the comment to "expose scheduling queue in factory.Config, instead of a function".

Huang-Wei commented:
@bsalamat @ravisantoshgudimetla PTAL. (I will do a final squash and remove the e2e test commit)

Huang-Wei committed:
ensure scheduler preemptor behaves in an efficient/correct path

- don't update nominatedMap cache when Pop() an element from activeQ
- instead, delete the nominated info from cache when it's "assumed"
- unit test behavior adjusted
- expose SchedulingQueue in factory.Config

Huang-Wei commented:
/retest

Huang-Wei commented:
/retest

Huang-Wei commented:
/retest

Huang-Wei commented:
Integration test keeps failing due to different flakes.

Huang-Wei commented:
/retes

Huang-Wei commented:
/retest

ravisantoshgudimetla reviewed (approved):
Approving the PR based on https://github.com/kubernetes/kubernetes/pull/70898#pullrequestreview-175971832

@Huang-Wei Thanks for identifying the issue and fixing it. Please prepare a PR for a test case as follow-up.

/lgtm
/approve

ravisantoshgudimetla commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -441,9 +446,8 @@ func (p *PriorityQueue) MoveAllToActiveQueue() {
+// NOTE: this function assumes lock has been acquired in caller
 func (p *PriorityQueue) movePodsToActiveQueue(pods []*v1.Pod) {
-	p.lock.Lock()
-	defer p.lock.Unlock()
```
.
